Dispatch initializers by their allocating entry point

jckarter opened the pull request:
- Change convenience initializers not to show up in vtables (unless `required` or `override`-ing designated initializers)
- Change convenience initializers only to have allocating entry points
- Change all initializers to dispatch in vtables via their allocating entry points, since the initializing entry point is only needed for `super.init` chains, and those can always be statically resolved

rdar://problem/29634243

jckarter committed:
SILGen: Emit convenience initializers as allocating entry points.

And only dispatch designated inits by their allocating entry points. rdar://problem/29634243

jckarter committed:
SIL: Don't include convenience initializers in vtables.

They should never be dynamically dispatched (unless `required`), so this entry should never be used.
We were accidentally dynamically dispatching to them in convenience-to-convenience `self.init`
delegations; fix that.

jckarter commented:
@swift-ci Please test source compatibility

slavapestov commented on lib/SILGen/SILGenApply.cpp:
```diff
@@ -805,46 +811,83 @@ class SILGenApply : public Lowering::ExprVisitor<SILGenApply> {
-        loc, selfMeta.getValue(), SGF.SGM.getLoweredType(objcMetaType)));
+    SILValue convertedValue;
+    switch (metatypeRepPair(givenMetatypeRep, destMetatypeRep)) {
+    case metatypeRepPair(MetatypeRepresentation::Thick,
```
This is cute, but I don't think its more readable than just an if/else if chain :)

slavapestov commented on lib/SILGen/SILGenBridging.cpp:
```diff
@@ -1377,6 +1377,20 @@ static SILFunctionType *emitObjCThunkArguments(SILGenFunction &SGF,
   SILDeclRef native = thunk.asForeign(false);
+
+  // If we're calling a native non-designated class initializer, we have to
+  // throw the `self` object we were given away to make a new one, since
```
To avoid confusion, maybe say 'deallocate' instead of 'throw'

jrose-apple commented on lib/AST/Decl.cpp:
```diff
@@ -5246,6 +5246,18 @@ static bool requiresNewVTableEntry(const AbstractFunctionDecl *decl) {
   if (decl->isFinal() || decl->isDynamic() || decl->hasClangNode())
     return false;
+  
+  // Initializers are not normally inherited, but required initializers can
```
I'm confused by this "inherited". Requiring a new vtable entry also happens when the type changes.

jckarter replied:
A convenience initializer is always effectively `final`, though, so even if it overrides a designated init while changing type, it'll override an existing vtable entry, but never need a new one of its own.

jrose-apple replied:
Ah, sure, but this comment is about all initializers. I would have expected something like "Only designated initializers and `required` convenience initializers appear in the vtable; bail out if this is an initializer that is neither designated nor `required`."

jrose-apple commented on lib/SILGen/SILGenApply.cpp:
```diff
@@ -1226,19 +1270,15 @@ class SILGenApply : public Lowering::ExprVisitor<SILGenApply> {
              && "some new kind of init context we haven't implemented");
-      useAllocatingCtor = static_cast<bool>(SGF.AllocatorMetatype) &&
-                          !ctorRef->getDecl()->isObjC();
+      useAllocatingCtor = ctorRef->getDecl()->isImportAsMember()
```
This is a different test. What's special about `isImportAsMember` that IRGen needs to know about it? What's wrong with `isFactoryInit()`?

jrose-apple commented on lib/SILGen/SILGenConstructor.cpp:
```diff
@@ -485,7 +481,7 @@ void SILGenFunction::emitClassConstructorAllocator(ConstructorDecl *ctor) {
-  if (ctor->isConvenienceInit() || ctor->hasClangNode()) {
+  if (ctor->hasClangNode()) {
     // For a convenience initializer or an initializer synthesized
     // for an Objective-C class, allocate using the metatype.
```
Comment needs updating here too, right?

jrose-apple reviewed (commented):
I think you'll still need some work for MissingMemberDecl (it will get simpler), but this is looking pretty reasonable to me. I'm not the SIL expert, though.

jckarter committed:
SILOptimizer: Turn type(of: self) from uninitialized self stack space into the self argument.

Before we changed convenience inits into allocating entry points, we allowed type(of: self) to be invoked on the uninitialized object, which was fine. This no longer Just Works when self doesn't even exist before `self.init` is called, but to maintain the old semantics and source compatibility, we can still just use the metatype value we were passed in.

jckarter commented:
@swift-ci Please test

jckarter commented:
@jrose-apple @slavapestov mind taking another look over this now? It should be close to ready to commit.

jckarter commented:
@swift-ci Please test

jrose-apple reviewed (commented):
Implementation looks good, some concerns about the test updates (of which there were a lot!). Also, maybe I scrolled too fast, but is there a test for the new `@objc` convenience initializer thunk, the one that does a `dealloc_partial_ref`?

jckarter commented on test/SILOptimizer/definite_init_diagnostics.swift:
```diff
@@ -432,18 +432,18 @@ class DelegatingCtorClass {
-    _ = ivar       // expected-error {{'self' used in property access 'ivar' before 'self.init' call}}
-    ivar = x       // expected-error {{'self' used in property access 'ivar' before 'self.init' call}}
+    _ = ivar       // expected-error {{'self' used before 'self.init' call}}
+    ivar = x       // expected-error* {{'self' used before 'self.init' call}}
```
Definite initialization for convenience inits now uses the same code path as for delegating value type initializers, so some of the diagnostics specialized for partially-initialized class instances no longer fire. I can look into the QoI regressions after, but they're consistent with what you get for a value type delegating initializer.

jrose-apple replied:
That seems wrong to me, then. A value type initializer is allowed to initialize a few fields and then throw those values away, but a convenience initializer isn't.

Either way, a test case should never use `*` on  an expected-error it's actually trying to test. Either the error is present or it's not.

jckarter replied:
I agree; like I said, I'd like to look into the diagnostic issues once the codegen changes are in, because they also affect value type initializers today.

jrose-apple replied:
I'm not happy with this, because this isn't just a diagnostic issue; it's a language change. Convenience initializers inherently *can't* do everything value initializers can. But I see the point about not having a long-lived branch that changes a massive number of tests.

jckarter replied:
There's no language change, only the diagnostics changed.

jckarter replied:
To be specific, "initialize a few fields and then throw it away" still does not work on convenience initializers because accessing any field of a class requires indirection through the instance pointer, and that pointer now does not exist until a self.init call happens.

jckarter commented on test/IRGen/ivar_destroyer.sil:
```diff
@@ -45,12 +44,10 @@ class NonTrivialDerived : TrivialBase {
 sil_vtable NonTrivialDerived {
-  #NonTrivialDerived.init!initializer.1: @init_NonTrivialDerived [override]
-  #NonTrivialDerived.init!allocator.1: @alloc_NonTrivialDerived
+  #TrivialBase.init!allocator.1: @alloc_NonTrivialDerived [override]
```
We didn't correctly follow override chains for initializing entry points, and by the previous rules, the allocating entry point wasn't established until the subclass made it `required`.

jckarter commented on lib/SILOptimizer/Mandatory/DIMemoryUseCollectorOwnership.h:
```diff
@@ -226,7 +230,11 @@ enum DIUseKind {
+  
+  /// This instruction is a load that's only used to answer a `type(of: self)`
+  /// question.
+  LoadForTypeOfSelf,
```
This is necessary for existing tests (as well as swiftpm and the compatibility suite) to keep working.

jckarter commented on test/SILOptimizer/definite_init_failable_initializers_diagnostics.swift:
```diff
@@ -128,18 +128,17 @@ class ErrantClass : ErrantBaseClass {
-      something(x) // expected-error {{'self' used inside 'catch' block reachable from self.init call}}
-      something(self.x) // expected-error {{'self' used inside 'catch' block reachable from self.init call}}
+      something(x) // expected-error* {{'self' used before 'self.init'}}
+      something(self.x) // expected-error* {{'self' used before 'self.init'}}
```
There's special handling for `catch` blocks trying to access partially-initialized class instances, but it looks irrelevant when there either is an object or isn't. This is again because convenience initializers now get checked exactly the same as other delegating initializers.

jckarter commented:
@swift-ci Please clean test

jckarter commented:
@swift-ci Please test source compatibility

jckarter commented:
@swift-ci Please test

jckarter committed:
Update tests based on review feedback.

jckarter commented:
@swift-ci Please test

jckarter commented:
@swift-ci Please test

jckarter commented:
@swift-ci Please test source compatibility

jckarter commented:
@swift-ci Please test source compatibility

jckarter commented:
@jrose-apple I think I got all your suggested updates to the tests. I ended up ripping out the changes to convenience overriding checking in the AST, because they had some fallout in the compatibility suite, and it didn't really end up simplifying the vtable formation rules because there were a lot of other things like PrintAsObjC, selector collision checking, and QoI for attempted overrides, that depended on the overrides being present in the AST. How's this look now?

jrose-apple commented on test/SILGen/objc_factory_init.swift:
```diff
@@ -42,58 +38,34 @@ extension Hive {
-  // CHECK:   throw [[ERROR]]
-  // CHECK: } // end sil function '$SSo4HiveC17objc_factory_initE15otherFlakyQueenABSo3BeeC_tKcfC'
-  convenience init(otherFlakyQueen other: Bee) throws {
-    try self.init(flakyQueen: other)
```
What happened to this test?

jckarter replied:
I judged that it isn't testing anything unique at this point, since the convenience init behavior is consistent now with other tests for failable delegating init behavior, so I removed it.

jrose-apple commented on lib/SILOptimizer/Mandatory/DIMemoryUseCollectorOwnership.cpp:
```diff
@@ -1519,13 +1519,48 @@ collectValueTypeDelegatingInitUses(const DIMemoryObjectInfo &TheMemory,
+    if (isa<LoadInst>(User)) {
+      auto UserVal = cast<SingleValueInstruction>(User);
+      if (UserVal->hasOneUse()
+          && isa<ValueMetatypeInst>(UserVal->getSingleUse()->get())) {
```
Hm. What happens if someone uses `type(of: self)` twice?

jckarter replied:
Each one gets its own load, at least with the way SILGen is currently implemented.

jrose-apple replied:
That seems worth having a test for, at least, in case SILGen ever figures out how peephole the load but not the ValueMetatypeInst.

jrose-apple commented on lib/SILOptimizer/Mandatory/DefiniteInitialization.cpp:
```diff
@@ -826,6 +842,28 @@ void LifetimeChecker::handleLoadUse(unsigned UseID) {
+    }
+    assert(valueMetatype);
+    auto metatypeArgument = load->getFunction()->getEntryBlock()->getArguments()
+      .back();
```
Nitpick: `load->getFunction()->getSelfMetadataArgument()`.

jrose-apple reviewed (commented):
Still a few comments but they're pretty small!

jrose-apple commented:
The override-in-AST thing is interesting and kind of worrisome but I totally agree that it shouldn't block your progress now.
